bin/runtime-common/src/messages_xcm_extension.rs

```diff
@@ -0,0 +1,170 @@
+// Copyright 2022 Parity Technologies (UK) Ltd.
+// This file is part of Cumulus.
+
+// Cumulus is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Cumulus is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Cumulus.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Module provides utilities for easier XCM handling, e.g:
+//! [`XcmExecutor`] -> [`MessageSender`] -> <outbound message queue>
+//!												|
+//!											<relayer>
+//!												|
+//! [`XcmRouter`] <- [`MessageDispatch`] <- <inbound message queue>
+
+use bp_messages::{
+	source_chain::MessagesBridge,
+	target_chain::{DispatchMessage, MessageDispatch},
+	LaneId,
+};
+use bp_runtime::{messages::MessageDispatchResult, AccountIdOf, Chain};
+use codec::{Decode, Encode};
+use frame_support::{dispatch::Weight, CloneNoBound, EqNoBound, PartialEqNoBound};
+use scale_info::TypeInfo;
+use xcm_builder::{DispatchBlob, DispatchBlobError, HaulBlob, HaulBlobError};
+
+/// PLain "XCM" payload, which we transfer through bridge
+pub type XcmAsPlainPayload = sp_std::prelude::Vec<u8>;
+
+/// Message dispatch result type for single message
+#[derive(CloneNoBound, EqNoBound, PartialEqNoBound, Encode, Decode, Debug, TypeInfo)]
+pub enum XcmBlobMessageDispatchResult {
+	InvalidPayload,
+	Dispatched,
+	NotDispatched(#[codec(skip)] &'static str),
+}
+
+/// [`XcmBlobMessageDispatch`] is responsible for dispatching received messages
+pub struct XcmBlobMessageDispatch<SourceBridgeHubChain, TargetBridgeHubChain, DispatchBlob> {
+	_marker:
+		sp_std::marker::PhantomData<(SourceBridgeHubChain, TargetBridgeHubChain, DispatchBlob)>,
+}
+
+impl<SourceBridgeHubChain: Chain, TargetBridgeHubChain: Chain, BlobDispatcher: DispatchBlob>
+	MessageDispatch<AccountIdOf<SourceBridgeHubChain>>
+	for XcmBlobMessageDispatch<SourceBridgeHubChain, TargetBridgeHubChain, BlobDispatcher>
+{
+	type DispatchPayload = XcmAsPlainPayload;
+	type DispatchLevelResult = XcmBlobMessageDispatchResult;
+
+	fn dispatch_weight(_message: &mut DispatchMessage<Self::DispatchPayload>) -> Weight {
+		log::error!(
+			target: crate::LOG_TARGET_BRIDGE_DISPATCH,
+			"[XcmBlobMessageDispatch] TODO: change here to XCMv3 dispatch_weight with XcmExecutor - message: ?...?",
+		);
+		// TODO:check-parameter - setup weight?
```

Comment:
Can you, please, add issue reference to those TODOs?

Reply:
BTW - looking at the `DispatchBlob` trait, I can see that the `dispatch_blob` function `returns the unexpectable weight consumed by the dispatch`, but the return value is `Result<(),_>`. That's kinda strange.  Maybe it is designed to be called from `on_idle`? But anyway - we need the worst-case weight of this function and ideally the unspent weight.

Reply:
@svyatonik 
yes, well, I saw that long time ago, and reported in original xcm-v3 pr several time: 
https://github.com/paritytech/polkadot/pull/4097#discussion_r992821284

I have this still in my TODOs, so I will probably open a separate issue for that `dispatch_blob`